packages/concerto-core/lib/decoratorutil.js

```diff
@@ -0,0 +1,409 @@
+        if (declaration.key){
+            if (declaration.key.decorators){
+                const constructOptions = {
+                    declaration: declaration.name,
+                    mapElement: 'KEY'
+                };
+                extractionDictionary = DecoratorUtil.constructDCSDictionary(extractionDictionary, namespace, declaration.key.decorators, constructOptions);
+                if (removeDecoratorsFromModel){
+                    declaration.key.decorators = undefined;
+                }
+            }
+        }
+        if (declaration.value){
+            if (declaration.value.decorators){
+                const constructOptions = {
+                    declaration: declaration.name,
+                    mapElement: 'VALUE'
+                };
+                extractionDictionary = DecoratorUtil.constructDCSDictionary(extractionDictionary, namespace, declaration.value.decorators, constructOptions);
+                if (removeDecoratorsFromModel){
+                    declaration.value.decorators = undefined;
+                }
+            }
+        }
+        return {
+            declaration,
+            extractionDictionary
+        };
+    }
+    /**
+    * Process the declarations to extract the decorators.
+    *
+    * @param {Object} sourceDecl - The source AST of the model
+    * @param {Object} extractionDictionary - the collection of extracted decorators
+    * @param {boolean} removeDecoratorsFromModel - flag to decide whether to remove decorators or not
+    * @param {string} namespace - namespace of the model
+    * @returns {Object} - constructed DCS Dict and processed declarations ast
+    * @private
+    */
+    static processDeclarations(sourceDecl, extractionDictionary, removeDecoratorsFromModel, namespace){
+        const processedDecl = sourceDecl.map(decl => {
+            if (decl.decorators) {
+                const constructOptions = {
+                    declaration: decl.name,
+                };
+                extractionDictionary = DecoratorUtil.constructDCSDictionary(extractionDictionary, namespace, decl.decorators, constructOptions);
+            }
+            if (removeDecoratorsFromModel){
+                decl.decorators = undefined;
+            }
+            if (decl.$class.endsWith('.MapDeclaration')) {
+                const processedMapDeclData = this.processMapDeclaration(decl, extractionDictionary, removeDecoratorsFromModel, namespace);
+                decl = processedMapDeclData.declaration;
+                extractionDictionary=processedMapDeclData.extractionDictionary;
+            }
+            if (decl.properties) {
+                const processedPropertiesData = this.processProperties(decl.properties, decl.name, extractionDictionary, removeDecoratorsFromModel, namespace);
+                decl.properties = processedPropertiesData.processedProperties;
+                extractionDictionary = processedPropertiesData.extractionDictionary;
```

Comment:
These types of providing arguments and then reading the same arguments from the returned data begs for contained state. I would either create a class and keep the state as a private property, the state in this case is `extractionDictionary` and `properties`(maybe), or, I would keep these in a `TransformationContext` object, and pass that object around, instead of calling functions and reassigning them.

Reply:
I didn't realise that this whole thing was a class, I think you can convert everything into a `DecoratorTransformer` or `DecoratorExtractor` class and accept all the things you accept in `collectExtractedDecoratorsAndProcessedModels` in the constructor, then keep `exctractionDictionary` as an internal state. Once constructed, you can move the logic in `collectExtractedDecoratorsAndProcessedModels` into a method called, `transform` or `extract`.

That way you don't need to pass it around and you can make the class public, since it will work with any model manager, and you can make all these functions that are not being called elsewhere private functions.